### PR TITLE
feat: add retain old password support

### PR DESCRIPTION
**Pull request by Kiliandeca**
Hello :wave:, this PR introduce the possibility of retaining the old password when the password change as described in https://github.com/petoju/terraform-provider-mysql/issues/17.

I would like to add a test but I'm not very fluent in Go and unless I'm mistaken there are currently no test that check if the user password works from which I could copy/inspire. Any feedback or advice is very welcome :pray:

### CODE REVIEW

File: mysql/resource_user.go [hunk 1]
```diff
@@ -198,6 +212,14 @@ func CreateUser(ctx context.Context, d *schema.ResourceData, meta interface{}) d
 		}
 	}
 
+	retainPassword := d.Get("retain_old_password").(bool)
+	if retainPassword {
+		_, err := checkRetainCurrentPasswordSupport(ctx, meta)
+		if err != nil {
+			return diag.Errorf("%v", err)
```
Comment (petoju) [inline mysql/resource_user.go:219]: Please make error message more descriptive by saying, where it came from.
Go usually goes with "stacktrace-like" errors, where more and more errors are prepended to the error message. That makes debugging easier.

File: mysql/resource_user.go [hunk 2]
```diff
@@ -100,10 +101,23 @@ func resourceUser() *schema.Resource {
...
+
+			"retain_old_password": {
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
 		},
 	}
 }
 
+func checkRetainCurrentPasswordSupport(ctx context.Context, meta interface{}) (bool, error) {
```
Comment (petoju) [inline mysql/resource_user.go:113]: Consider returning just `bool` or just `error`.
It seems you never use the `bool` output, so maybe it is useless?

File: mysql/resource_user.go [hunk 3]
```diff
@@ -2,15 +2,16 @@ package mysql
```
Comment (petoju) [inline mysql/resource_user.go:2]: For tests:

implement your own check similar to this one
https://github.com/petoju/terraform-provider-mysql/blob/master/mysql/resource_user_test.go#L151-L181
that will check, whether one can authenticate as given user with given password. Maybe your test will just try to connect to MySQL with that data.

Then implement test like this:
https://github.com/petoju/terraform-provider-mysql/blob/master/mysql/resource_user_test.go#L54-L89
you don't have to run TestCheckResourceAttr. Without that, you'll run some written config, then possibly do some check. Next step will change password by specifying another config and run the check again to verify the old password stayed in place.
The last step would be to check, whether authentication with some random password works - that is a verification whether our tests make sense.

I also recommend doing test for the case one doesn't want to retain password.

**Review by petoju — CHANGES_REQUESTED** [review]

**Comment by Kiliandeca** [comment]
Thanks for your quick review and guidance on how to implement the tests. Let me know if there are still things you want me to improve.

**Review by petoju — APPROVED** [review]
Looks good!

**Comment by petoju** [comment]
Released in v3.0.42

**Pull request merged**